Complete the move to new serial library

danielb987 opened the pull request:
This is a continuation of PR #13163. It moves most of the remaining connections to the new serial library.
Maybe it's a candidate for the label `After Next Production Release`?

---

There is one connection that's remaining: `jmri.jmrix.serialsensor.SerialSensorAdapter`. That class uses listeners for the CTS, DSR and CD signals on the serial port.

There is a way to listen on these events in com.fazecast.jSerialComm.SerialPortDataListener, but there is a big BUT. Quote from [com.fazecast.jSerialComm.SerialPortDataListener](https://github.com/Fazecast/jSerialComm/blob/master/src/main/java/com/fazecast/jSerialComm/SerialPortDataListener.java), lines 61-64:

> It is recommended to **only** use the LISTENING_EVENT_DATA_AVAILABLE, LISTENING_EVENT_DATA_RECEIVED, LISTENING_EVENT_DATA_WRITTEN, and/or LISTENING_EVENT_PORT_DISCONNECTED listening events in production or cross-platform code since underlying differences and lack of support for the control line status and error events among the various operating systems and device drivers make it unlikely that code listening for these events will behave similarly across different serial devices or OS's, if it works at all.

So the question is what to do about it? Is this class used or can it be removed? I don't find any reference to it in the [documentation](https://www.jmri.org/help/en/html/hardware/index.shtml).

bobjacobsen commented:
`SerialSensor` was an attempt to use the serial port status leads for sensor inputs.  Although it did work with our RXTX library during testing, I don't know of anybody ever using it in practice. It never really worked right with the PJC library and nobody mentioned that.  

I think it can be removed.

danielb987 commented:
> `SerialSensor` was an attempt to use the serial port status leads for sensor inputs.

I like the idea. I have done something similar 20 years ago. It's a bit sad it doesn't work well with modern serial ports.

I have removed it now. Can we remove all the references to purejavacomm from AbstractSerialPortController or could it be any scripts or external software that are still using it? And how about the libraries themselves? Do we want to keep the purejavacomm libraries?

bobjacobsen commented:
I think we need to leave the purejavacomm libraries in place.  We've published scripts that use them, so it seems likely that there are users with their own copies of those scripts:

```
% grep -rl javacomm jython     
jython/serialinput/Sensor_Channel_To_JMRI/SensorSerialClose.py
jython/serialinput/Sensor_Channel_To_JMRI/Sensor_Scan.py
jython/serialinput/SerialPortDevice.py
jython/serialinput/Turnout_Channel_from_JMRI/LightDataTransfer.py
jython/serialinput/Turnout_Channel_from_JMRI/Light_Table_Build.py
jython/serialinput/Turnout_Channel_from_JMRI/TurnoutDataTransfer.py
jython/serialinput/Turnout_Channel_from_JMRI/Turnout_Table_Build.py
jython/serialinput/SerialPortThrottle.py
jython/SerialPortRFID.py
jython/ParallelPortTurnouts.py
jython/SerialPortTest.py
jython/ParallelPortTest.py
jython/ComPortEnumeration.py
```

There's no real harm in leaving the libraries in place.

bobjacobsen reviewed (dismissed):
Thanks for doing this!

danielb987 commented:
> Thanks for doing this!

Thank you!

I have updated AbstractSerialPortController so that it doesn't use PJC and I have updated ArchitectureTest to prevent any use of PJC. I have also added a note to ArchitectureTest that scripts uses PJC.

danielb987 commented:
This PR is ready for review.
Should it be included in the next test release or wait for after the next production release?

bobjacobsen commented:
Thanks for asking.  I think that it would be wise to keep this one until after the production release. It changes a few systems, and we don't have anybody specific to test those.  (This is different from PR #13163 where we had somebody with hardware who could test that specific change)

danielb987 commented:
> Thanks for asking. I think that it would be wise to keep this one until after the production release. It changes a few systems, and we don't have anybody specific to test those.

I agree. I have changed this PR to WIP to prevent accidental merge. Feel free to remove WIP when time come.
